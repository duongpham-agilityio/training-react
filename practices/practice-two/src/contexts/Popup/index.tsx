import { ReactNode, MouseEvent, useState, useMemo, useCallback } from 'react';

// Components
import { Popup } from 'components';

// Contexts
import { PopupContext } from 'contexts/Popup/context';

// Constants
import { MESSAGES } from '@constants';

// const initialState = {
//   isShown: false,
//   accepter: () => {},
// };

const PopupProvider = ({ children }: { children: ReactNode }) => {
  const [state, setState] = useState<{
    isShown: boolean;
    accepter: (_event: MouseEvent) => void;
  } | null>(null);

  const props = useMemo(() => {
    return {
      dispatch: (callback: () => void) => {
        setState({
          isShown: true,
          accepter: callback,
        });
      },
    };
  }, []);

  const onAccept = useCallback(
    (event: MouseEvent) => {
      state?.accepter(event);

      setState(null);
    },
    [state]
  );

  return (
    <PopupContext.Provider value={props}>
      {children}

      {state?.isShown && (
        <Popup
          title={MESSAGES.DELETE_TITLE}
          description={MESSAGES.DELETE_DESCRIPTION}
          onCancel={() => setState(null)}
          onAccept={onAccept}
        />
      )}
    </PopupContext.Provider>
  );
};

export default PopupProvider;
