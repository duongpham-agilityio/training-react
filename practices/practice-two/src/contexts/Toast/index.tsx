import { ReactNode, useCallback, useState } from 'react';

// Components
import { NotificationProps, Notification } from 'components';

// Context
import { ToastContext } from 'contexts/Toast/context';

const ToastProvider = ({ children }: { children: ReactNode }) => {
  const [notification, setNotification] = useState<
    NotificationProps | undefined
  >();

  const handleSetNotification = useCallback((props: NotificationProps) => {
    setNotification(props);

    setTimeout(() => {
      setNotification(undefined);
    }, 2000);
  }, []);

  return (
    <ToastContext.Provider value={{ setNotification: handleSetNotification }}>
      {children}

      {notification && <Notification {...notification} />}
    </ToastContext.Provider>
  );
};

export default ToastProvider;
