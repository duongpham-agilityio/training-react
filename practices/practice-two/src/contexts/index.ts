export { default as ToastProvider } from 'contexts/Toast';
export { default as PopupProvider } from 'contexts/Popup';
