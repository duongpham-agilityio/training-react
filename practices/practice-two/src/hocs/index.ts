export * from './withErrorBoundaries';
export * from './withUseForm';
