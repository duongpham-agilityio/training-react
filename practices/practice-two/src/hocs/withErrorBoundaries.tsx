import React, { FunctionComponent, Component } from 'react';

// Component
import { Error } from 'components';

export const withErrorBoundaries = <T extends object>(
  WrappedComponent: FunctionComponent<T>
) => {
  class Wrapper extends Component<T, { hashError: boolean }> {
    constructor(props: T) {
      super(props);
      this.state = {
        hashError: false,
      };
    }

    componentDidCatch(): void {
      this.setState({
        hashError: true,
      });
    }

    render(): React.ReactNode {
      if (this.state.hashError) {
        return <Error />;
      }

      return <WrappedComponent {...this.props} />;
    }
  }

  return Wrapper;
};
