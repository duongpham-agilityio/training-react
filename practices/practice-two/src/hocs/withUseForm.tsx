import { FC, Suspense, lazy, useCallback, useMemo, useState } from 'react';
import useMutation from 'swr/mutation';

// Hooks
import { FormData, useToastContext } from 'hooks';

// Components
const Form = lazy(() => import('components/Form'));

// Constants
import { ENDPOINT, MESSAGES } from '@constants';

// Helpers
import { axiosConfig } from 'helpers';

// Types
import { Book } from 'types';

type FormStatus = 'update' | 'create';
interface StateType {
  formData: Omit<Book, 'publishDate'> & {
    publishDate?: number;
  };
  title: string;
  type: FormStatus;
}

export interface WithUseFormProps {
  dispatchAction: (data: StateType) => void;
}

export const withUseForm = <T extends WithUseFormProps>(Component: FC<T>) => {
  const FormHOC = (props: Omit<T, keyof WithUseFormProps>) => {
    const { setNotification } = useToastContext();
    const [form, setForm] = useState<StateType | null>(null);
    const [formStatus, setFormStatus] = useState<FormStatus>('create');
    const { trigger: triggerAddBook } = useMutation(
      `${ENDPOINT.BOOKS}${ENDPOINT.SORT}`,
      (_, { arg }: { arg: Omit<FormData, 'imageName'> }) =>
        axiosConfig.post(ENDPOINT.BOOKS, arg)
    );
    const { trigger: triggerUpdateBook } = useMutation(
      `${ENDPOINT.BOOKS}/${form?.formData.id}`,
      (key: string, { arg }: { arg: FormData & { updatedAt: number } }) =>
        axiosConfig.patch(key, arg)
    );

    const handleCloseForm = useCallback(() => setForm(null), []);

    const dispatchAction = useCallback((data: StateType) => {
      setForm(data);
      setFormStatus(data.type);
    }, []);

    const onTriggerCreate = useCallback(
      (result: FormData) =>
        triggerAddBook(result)
          .then(() =>
            setNotification({
              message: MESSAGES.ADD_SUCCESS,
              title: MESSAGES.ADD_TITLE,
            })
          )
          .catch(() =>
            setNotification({
              message: MESSAGES.ERROR_TITLE,
              title: MESSAGES.EMPTY_FIELD,
              type: 'error',
            })
          )
          .finally(handleCloseForm),
      [handleCloseForm, setNotification, triggerAddBook]
    );

    const onTriggerUpdate = useCallback(
      (result: FormData) => {
        const { author, description, imageURL, name, publishDate } = result;
        const payload: FormData & { updatedAt: number } = {
          author,
          description,
          imageURL,
          name,
          updatedAt: new Date().getTime(),
          publishDate,
        };

        return triggerUpdateBook(payload)
          .then(() =>
            setNotification({
              message: MESSAGES.UPDATE_TITLE,
              title: MESSAGES.UPDATE_SUCCESS,
            })
          )
          .catch(() =>
            setNotification({
              message: MESSAGES.ERROR_TITLE,
              title: MESSAGES.EMPTY_FIELD,
              type: 'error',
            })
          )
          .finally(handleCloseForm);
      },
      [handleCloseForm, setNotification, triggerUpdateBook]
    );

    const onSubmit = useMemo(() => {
      const handler = {
        create: onTriggerCreate,
        update: onTriggerUpdate,
      };

      return handler[formStatus];
    }, [formStatus, onTriggerCreate, onTriggerUpdate]);

    return (
      <>
        <Component {...(props as T)} {...{ dispatchAction }} />
        {form && (
          <Suspense>
            <Form
              value={form.formData}
              title={form.title}
              type={form.type}
              onClose={handleCloseForm}
              onSubmit={onSubmit}
            />
          </Suspense>
        )}
      </>
    );
  };

  return FormHOC;
};
