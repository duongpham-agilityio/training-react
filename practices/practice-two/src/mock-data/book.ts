import { Book } from 'types';

const title = `HTML/CSS book name will displayed here, test with a
long text, will display full`;

const description = `Will show full description here, with a long text, will limit about 200  character.
Then will display the show more/show less for it. You   with a long text, will limit about 200  character.
Then will display the show more/show less for it. You  with a long text, will limit about 200  character.
Then will display the show more/show less for it. You  with a long text, will limit about 200  character.
Then will display the show more/show less for it. You  with a long text, will limit about 200  character.
Then will display the show more/show less for it. You can see this ex more/show less for it. You can see this ex more/show less for it. You can see this ex You can see this ex You can see this ex more/show less for it. You can see this ex`;

const author = 'Cu.Nguyen';
const imageURL =
  'http://books.google.com/books/content?id=KzzXzqLzXi8C&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api';

export const book: Book = {
  name: title,
  author,
  description,
  imageURL,
  id: 0,
  publishDate: new Date().getTime(),
  createdAt: new Date().getTime(),
  updatedAt: new Date().getTime(),
  deletedAt: null,
};
