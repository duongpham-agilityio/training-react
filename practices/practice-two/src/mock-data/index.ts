export * from 'mock-data/book';
export * from 'mock-data/books';
