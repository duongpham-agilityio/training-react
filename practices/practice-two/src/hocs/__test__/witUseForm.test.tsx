// Mocks
import 'components/__test__/mocks/image.test';

import { fireEvent, render, waitFor } from '@testing-library/react';
import { books } from 'mock-data';

// HOCs
import { WithUseFormProps, withUseForm } from 'hocs/withUseForm';
import { ReactNode } from 'react';
import { ToastProvider } from 'contexts';

const Component = ({ dispatchAction }: WithUseFormProps) => {
  return (
    <>
      <button
        onClick={() => {
          dispatchAction({
            formData: books[0],
            title: 'Form',
            type: 'create',
          });
        }}
        data-testid="create-btn"
      >
        Create
      </button>
      <button
        onClick={() => {
          dispatchAction({
            formData: books[0],
            title: 'Form',
            type: 'update',
          });
        }}
        data-testid="update-btn"
      >
        Update
      </button>
    </>
  );
};

const MockComponent = withUseForm(Component);
const Wrapper = ({ children }: { children: ReactNode }) => (
  <ToastProvider>{children}</ToastProvider>
);

const setup = () =>
  render(<MockComponent />, {
    wrapper: Wrapper,
  });

describe('withUseForm', () => {
  it('Match to snapshot', () => {
    const { container } = setup();

    expect(container).toMatchSnapshot();
  });

  it('Create action', async () => {
    const { getByTestId, container } = setup();

    fireEvent.click(getByTestId('create-btn'));

    await waitFor(() => expect(container).toMatchSnapshot());
  });
});
