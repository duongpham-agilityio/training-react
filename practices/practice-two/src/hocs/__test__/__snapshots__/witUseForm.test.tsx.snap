// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withUseForm Create action 1`] = `
<div>
  <button
    data-testid="create-btn"
  >
    Create
  </button>
  <button
    data-testid="update-btn"
  >
    Update
  </button>
</div>
`;

exports[`withUseForm Match to snapshot 1`] = `
<div>
  <button
    data-testid="create-btn"
  >
    Create
  </button>
  <button
    data-testid="update-btn"
  >
    Update
  </button>
</div>
`;
